Add more vibe animations & clean up folders

matthewbeck opened the pull request:
Added animations
• Pulse
• Sparkle
• Fireworks
• Improved cursor spin
• Set foundation for calling random animation from array

matthewbeck committed:
Add more vibe animations & clean up folders

Added animations
• Pulse
• Sparkle
• Firework
• Improved cursor spin
• Set foundation for calling random animation from array

matthewbeck commented:
Big push with a lot of code. Might need a hand debugging vibe cursor spin. Every once in a while, it gets stuck as it loops from LED 180 to LED 0

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+void rainbow_vibe();
+void vibe_cursor_spin();
+void vibe_pulse();
+void vibe_sparkle();
```
I'm not sure these are needed, though they don't hurt

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+//Store animations in array
+void (*animationFunctions[])() = {rainbow_vibe, vibe_cursor_spin, vibe_pulse, vibe_sparkle};
+
+int numAnimations = sizeof(animationFunctions) / sizeof(animationFunctions[0]);
```
😍

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+{
+  
+  Serial.begin(115200);
+  delay(1000); // Put a small delay to ensure everything is setup properly
```
100-300 milliseconds ought to be enough.

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+  fill_solid(leds, NUM_LEDS, 0x000000); //turn all leds off
+  FastLED.show();
+  
+  FastLED.setBrightness(100); //Set Brightness of LED strip
```
Great seeing this here to reduce wattage while testing. Once moved to final hardware make sure to bump the brightness up

matthewbeck committed:
Update the way to call animation functions + animation improvements

• Change way we call animations to being an array that we cycle through. Have made some functions (rainbow) rarer than others
• Enhancements to some animations that weren't as tight as I wanted them.
• Fireworks animation still isn't where I want it to be, but getting closer

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+// 1. Vibe Flash animation
+//// Animation outputs colour on all LEDS when button is pressed. Sets all LEDs to black when button is depressed
+///////////////////////////////////////
+void vibe_flash()
+{
+    if(is_button_pressed) {
+      for(int i = 0; i < NUM_LEDS; i++){
+		uint8_t bri = 255;
+		leds[i] = CHSV(global_hue, 200, bri);
+	}
+  }
+    else {
+      for(int i = 0; i < NUM_LEDS; i++){
+      leds[i] = CHSV(0, 0, 0);
+      } //If no button press, LEDs off
+    }
+    FastLED.show();
+  
```
```suggestion
void vibe_flash()
{
    for(int i = 0; i < NUM_LEDS; i++){
	uint8_t bri = (is_button_pressed) ? 255 : 0;
	leds[i] = CHSV(global_hue, 200, bri);
    }
    FastLED.show();
  
```

Calculate the value of `bri` based on whether `is_button_pressed`.

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+{
+  // a colored dot moving around a cicle, with fading trails
+  if(is_button_pressed){
+  fadeToBlackBy( leds, NUM_LEDS, 3);
```
This is clever. Never thought of doing it that way.

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+  FastLED.show();
+  
+if (pos >= 255){
+  pos = 0;
```
There is a bug here:

`leds[pos] += CHSV( global_hue, 255, 255);`

The size of the `leds` array is NUM_LEDS, however `pos` could potentially be between [0,255]. When then the value is written to leds[255] this leads to memory corruption. 

1) make sure that data is never written to the `leds` array beyond index `NUM_LEDS-1`.

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+  if(is_button_pressed == true & last_buttonpress == true){
+  pos++;
+  
+  }
```
Careful here: Incrementing the value after you've checked that the value is in bounds, means that the value could be bad when you go to use it next. 

Eg. lets say there is a value X which we only want to be between [0,3]

```
print(X) => 2
check(X) => 2
Inc(X) => 3
// Loop ----------
print(X) => 3
check(X) => 3
Inc(X) => 4
// loop ----------
print(X) => 4.  // Value is used out of bounds
check(X) => 0
Inc(X) => 1.  // Value 0 is never used becuase it is incremented immediately.
```

It's better to increment first, then check.
```
pos++;
if (pos >= NUM_LEDS){
  pos = 0;
}
```

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+void trunk_charge(){
+  // Charge up the trunk with a green light
+  if(is_button_pressed){
+  leds[pos] = CHSV(96, 237, 255);
```
Interesting Saturation value here(237) curious why that one? is this color matching a color palette someonewhere? If so leave a comment with where to find it so other contributors don't much with the color matching

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+      }
+      bri = bri - 5;
+    }
+    //turn lights off after 2 seconds have passed since firework launch
+    else if(millis() - flashing_start_time >= 2000) {
+     for(int i = 0; i < NUM_LEDS; i++){
+    leds[i] = CHSV(0, 0, 0);
+    }
+    flashing = false;
+    }
+    }
+    else {
```
This is particularly hard to read. could use an `auto format`

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+      for(int i = 0; i < NUM_LEDS; i++){
+      leds[i] = CHSV(global_hue, 200, bri);
+      }
+      bri = bri - 5;
```
Careful these are using two different values of `bri`.

line 265 is using the variable defined on 264.
line 271 is using the variable defined on line 49.

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+  last_buttonpress = is_button_pressed;
+}
+
+void addGlitter( fract8 chanceOfGlitter) 
+{
+  if( random8() < chanceOfGlitter) {
+    leds[ random16(NUM_LEDS) ] += CRGB::White;
+  }
+}
```
Might be helpful to add a reference to where this code came from so you can find it later

jazzz commented on animation/animation.ino:
```diff
@@ -0,0 +1,394 @@
+    draw();
+    last_draw = millis();
+  }
+}
```
👨‍🍳
